### ISSUE TITLE
check_mod_item_types query times regressed bigly

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#65147]
There is a crate inside of Fuchsia whose compile times increased from ~500ms to ~120s in the last month of Rust compiler updates. It hasn't been easy so far to provide a minimal example, but I will keep working at it.

Some bisecting revealed that it actually consisted of two performance regressions, both affecting the time spent in `check_mod_item_types`:

| Commit hash                              | Item                                     | Self time | % of total time | Item count | Cache hits | Blocked time | Incremental load time |
|------------------------------------------|------------------------------------------|-----------|-----------------|------------|------------|--------------|-----------------------|
| 87b0c9036f0557568cfc939542508e9adf874b83 | check_mod_item_types                     | 34.92ms   | 7.120           | 6          | 0          | 0.00ns       | 0.00ns                |
| 34e82a7b793a6cdd27df762bf46bab8cdc92b14a | check_mod_item_types                     | 21.92s    | 97.786          | 6          | 0          | 0.00ns       | 0.00ns                |
| 66bf391c3aabfc77f5f7139fc9e6944f995d574e | check_mod_item_types                     | 21.76s    | 97.838          | 6          | 0          | 0.00ns       | 0.00ns                |
| 0221e265621a5fcc68ca62bdcdeabad1882a0e9a | check_mod_item_types                     | 133.53s   | 99.526          | 6          | 0          | 0.00ns       | 0.00ns                |

The first regression occurred in the rollup https://github.com/rust-lang/rust/pull/64354, and the second occurred sometime in the two weeks between changes 66bf391c3aabfc77f5f7139fc9e6944f995d574e..0221e265621a5fcc68ca62bdcdeabad1882a0e9a.

There is heavy use of async/await in the code, and I've identified a couple places where taking out `.await` makes the compile finish right away. This makes me suspect https://github.com/rust-lang/rust/pull/64292 for the first regression, but I'm not sure.

I'll update the issue as I have more information.

**Comment by tmandry** [comment]
Interestingly, removing `.await` from _either_ of these places makes the compile much faster. If I leave both in, it's slow, but taking one or both out makes it faster. I suspect there's some kind of cycle being created when both are left in.

**Comment by tmandry** [comment]
I don't have a lot more information to put here, except that the compile time seems to increase with the size of a particular call graph of async fns (it's not binary slow/fast, but different gradations of slow). This continues to make it hard to minimize.

I did some profiling, and noticed a lot of time being spent collecting items in `SmallVec::iter`. Unfortunately I couldn't see what was calling that function. There was some speculation that it could be from [this code](https://github.com/rust-lang/rust/blob/f3c9cece7b6829e6fd7854a1aee6a1619a81a38c/src/librustc_typeck/check/mod.rs#L1570-L1573).

See [this chat thread](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/perf.20callers) for more details, and some flame graphs of the profiles I ran.
